scenario.py

```diff
@@ -189,33 +192,172 @@ def __init__(self, params, experiment_path):
+        partners_with_shared_clusters.sort(key=operator.attrgetter("cluster_count"), reverse=True)
+        partners_with_specific_clusters.sort(key=operator.attrgetter("cluster_count"), reverse=True)
+
+        # Compose the list of different labels in the dataset
+        labels = list(set(y_train))
+        random.seed(42)
+        random.shuffle(labels)
+
+        # Check coherence of the split option:
+        nb_diff_labels = len(labels)
+        specific_clusters_count = sum([p.cluster_count for p in partners_with_specific_clusters])
+        if partners_with_shared_clusters:
+            shared_clusters_count = max([p.cluster_count for p in partners_with_shared_clusters])
+        else:
+            shared_clusters_count = 0
+        assert specific_clusters_count + shared_clusters_count <= nb_diff_labels
+
+        # Stratify the dataset into clusters per labels
+        x_train_for_cluster, y_train_for_cluster, count_per_cluster = {}, {}, {}
+        for label in labels:
+            idx_in_full_trainset = np.where(y_train == label)
+            x_train_for_cluster[label] = x_train[idx_in_full_trainset]
+            y_train_for_cluster[label] = y_train[idx_in_full_trainset]
+            count_per_cluster[label] = len(y_train_for_cluster[label])
+
+        # For each partner compose the list of clusters from which they will draw data samples
+        index = 0
+        for p in partners_with_specific_clusters:
+            p.clusters_list = labels[index:index + p.cluster_count]
+            index += p.cluster_count
+
+        shared_clusters = labels[index:index + shared_clusters_count]
+        for p in partners_with_shared_clusters:
+            p.clusters_list = random.sample(shared_clusters, k=p.cluster_count)
+
+        # We need to enforce the relative data amounts configured.
+        # It might not be possible to distribute all data samples, depending on...
+        # ... the coherence of the relative data amounts and the split option.
+        # We will compute a resize factor to determine the total nb of samples to be distributed per partner
+
+        # For partners getting data samples from specific clusters...
+        # ... compare the nb of available samples vs. the nb of samples initially configured
+        resize_factor_specific = 1
+        for p in partners_with_specific_clusters:
+            nb_available_samples = sum([count_per_cluster[cl] for cl in p.clusters_list])
+            nb_samples_configured = int(amounts_per_partner[p.id] * len(y_train))
+            ratio = nb_available_samples / nb_samples_configured
+            resize_factor_specific = min(resize_factor_specific, ratio)
+
+        # For each partner getting data samples from shared clusters:
+        # ... compute the nb of samples initially configured and resize it,
+        # ... then sum per cluster how many samples are needed.
+        # Then, find if a cluster is requested more samples than it got, and if yes by which factor
```

Comment:
find if a cluster has requested more samples than it has, and if yes by which factor

Reply:
Ok